File: HackerRank/WeekOfCode/WeekOfCode34/MaximumGcdAndSum.cpp
#include <bits/stdc++.h>
using namespace std;

int A[1000001];
int B[1000001];
char C[1000001];
int n;

int maximumGcdAndSum() {

    int max = 1000001;
    
    for (int i=0; i<n; i++)
    {
        C[A[i]] |= 1;
        C[B[i]] |= 2;
    }

    int maxsum = 0;
    for (int i=max-1; i>=1; i--)
    {
        int maxa = 0;
        int maxb = 0;
        for (int j=i; j<max ;j+=i)
        {
            int v = C[j];
            if ((v & 1)!=0) { maxa=j; }
            if ((v & 2)!=0) { maxb=j; }
        }
        if (maxa!=0 && maxb!=0)
        {
            maxsum = maxa + maxb;
            break;
        }
    }
    
    return maxsum;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(0);
    cin >> n;
    for (int i = 0; i < n; i++) {
       cin >> A[i];
    }
    for (int i = 0; i < n; i++) {
       cin >> B[i];
    }
    int res = maximumGcdAndSum();
    cout << res;
    return 0;
}

File: HackerRank/WeekOfCode/WeekOfCode34/MaximumGcdAndSum.cs
namespace HackerRank.WeekOfCode34.MaximumGcdAndSum
{
	using System;
	using System.Collections.Generic;
	using System.IO;
	using System.Linq;
	using static FastIO;

	class Solution
	{

		static int maximumGcdAndSum(int[] A, int[] B)
		{

			var table = new byte[1000001];
			foreach (var a in A) table[a] |= 1;
			foreach (var b in B) table[b] |= 2;

			int maxsum = 0;
			for (int i = table.Length - 1; i >= 1; i--)
			{
				int maxa = 0;
				int maxb = 0;
				for (int j = i; j < table.Length; j += i)
				{
					var v = table[j];
					if ((v & 1) != 0) maxa = j;
					if ((v & 2) != 0) maxb = j;
				}

				if (maxa != 0 && maxb != 0)
				{
					maxsum = maxa + maxb;
					break;
				}
			}

			return maxsum;
		}

		static void Main(String[] args)
		{
			int n = Ni();
			int[] A = Ni(n);
			int[] B = Ni(n);
			int res = maximumGcdAndSum(A, B);
			Console.WriteLine(res);
		}
	}



}